File: output/lec-python-v2/exercises/10-oop-多态契约/practice04.py
"""
[难度: ⭐⭐⭐]
[所属知识点: @abstractmethod 强制契约]
[预计完成时间: 20 分钟]

题目描述:
基于 practice03 的 Payment 抽象基类,
定义 Alipay 和 WeChatPay 子类,实现 execute。
验证:子类能正常实例化,漏写 execute 则报错。

示例:
    >>> alipay = Alipay()
    >>> alipay.execute(99.0)
    支付宝支付 99.0 元
    >>> class BrokenPay(Payment): pass
    >>> BrokenPay()  # 报 TypeError
"""

# ======================
# 学员代码区(以 pass 作为占位符)
# ======================
import abc

class Payment(abc.ABC):
    @abc.abstractmethod
    def execute(self, amount):
        ...

class Alipay(Payment):
    pass  # 请实现 execute

class WeChatPay(Payment):
    pass  # 请实现 execute

# ======================
# 测试区(教师可复制到终端验证)
# ======================
if __name__ == '__main__':
    # 测试 1:Alipay().execute(99.0) 正确
    # 测试 2:WeChatPay().execute(50.0) 正确
    # 测试 3:class BrokenPay(Payment): pass → BrokenPay() 报错
    pass

File: output/iter-R9-js-build-pedagogy/course/lesson03/in_class/practice04.js
/*
[难度: ★★]
[所属知识点: 反模式识别(NCDL)]
[预计完成时间: 10 分钟]

题目:下面代码有 2 个反模式,找出来。

    function checkout(cartTotal, payment) {
        if (payment.type === "alipay") {
            return alipayPay(cartTotal);
        } else if (payment.type === "wechat") {
            return wechatPay(cartTotal);
        }
    }

反模式提示:消费者门控的两个约束
*/

if (require.main === module) {
  console.log("2 个反模式:");
  console.log("  1. 用 if/elif 判断 payment.type → 违反'不能判断类型'约束");
  console.log("  2. 函数体超过 4 行 → 违反'消费者函数 ≤4 行'约束");
  console.log("");
  console.log("修正后:");
  console.log("  function checkout(cartTotal, payment) {");
  console.log("      return payment.execute(cartTotal);  // 多态调用,不判断类型");
  console.log("  }");
}
